tests/test_local_move.py

```diff
@@ -0,0 +1,181 @@
+        return target_logpdf_fxn(x_full)  # return restrained_logpdf_fxn(x_full)
+
+    # apply any valid MCMC move to this subproblem
+    x_sub = x[selection_mask]
+    x_next_sub, aux = mcmc_move(x_sub, subproblem_logpdf)
+    x_next = x.at[selection_mask].set(x_next_sub)
+
+    return x_next, aux
+
+
+def test_ideal_gas():
+    """Run HMC on subsets of an ideal gas system, where the subsets are selected based on a geometric criterion.
+    Assert that an observable based on this criterion doesn't drift after a large number of updates when using
+    local_resampling_move, and assert that it does drift when using an ablated version of local_resampling_move."""
+    np.random.seed(2022)
+
+    # make 2D ideal gas system
+    box_size = 5
+    dim = 2
+
+    box = np.eye(dim) * box_size
+    n_particles = 1000
+
+    def ideal_gas_2d_logpdf_fxn(x):
+        return 0.0
+
+    x0 = np.random.rand(n_particles, 2) * box_size
+
+    # make function that preferentially selects particles near center
+    center = np.ones(dim) * (box_size / 2)
+    r0 = box_size / 6
+
+    def central_particle_selection_log_prob_fxn(x):
+        distance_from_center = lambda x_i: jnp.linalg.norm(delta_r(x_i, center, box))
+        r = vmap(distance_from_center)(x)
+
+        return jnp.where(r > r0, -10 * (r - r0) ** 4, 0.0)
+
+    # define any correct MCMC move fxn -- e.g. a composition of other correct MCMC moves
+    def run_multiple_hmc_moves(x, logpdf_fxn, dt=1e-4, n_steps=100, n_moves=100):
+        hmc_move = make_hmc_mover(x, logpdf_fxn, dt=dt, n_steps=n_steps)
+        log_accept_probs = []
+        for _ in range(n_moves):
+            x, log_accept_prob = hmc_move(x)
+            log_accept_probs.append(log_accept_prob)
+
+        return x, np.array(log_accept_probs)
+
+    # define correct and incorrect version of local move
+    # (with the same target, same particle selection method, and same MCMC move)
+    common_kwargs = dict(
+        target_logpdf_fxn=ideal_gas_2d_logpdf_fxn,
+        particle_selection_log_prob_fxn=central_particle_selection_log_prob_fxn,
+        mcmc_move=run_multiple_hmc_moves,
+    )
+    correct_local_move = partial(local_resampling_move, **common_kwargs)
+    incorrect_local_move = partial(naive_local_resampling_move, **common_kwargs)
+
+    # test that num particles near center doesn't change dramatically
+    def particle_frac_near_center(x):
+        return np.mean(np.exp(central_particle_selection_log_prob_fxn(x)))
+
+    def assert_correctness(local_move):
+        # expect no drift
+        traj, aux_traj = expect_no_drift(
+            x0, local_move, observable_fxn=particle_frac_near_center, n_local_resampling_iterations=100
+        )
+
+        # assert move was not trivial
+        avg_accept_prob = np.mean([np.mean(np.exp(log_accept_probs)) for log_accept_probs in aux_traj])
+        assert avg_accept_prob > 0.1
+        assert np.max(np.abs(traj[0] - traj[-1])) > r0
```

Comment:
what is the purpose of this test? this seems to check that the first last frame has at least one particle that moved more than `r0`?

Reply:
also, `np.abs(x0 - x1)` i think returns maximum displacement along any dimension `(x,y,z)` and not the actual distance.

Reply:
To assert that the move was not trivial.

Reply:
Is there a more stringent thing that would be easy to assert?

Reply:
(Later assertions may pass trivially if `traj[t]` is constant as a function of `t`)

Reply:
hm not sure either, maybe check that at least `x` number of particles have moved outside of `r0` (as opposed to just 1 particle)

Reply:
Anyways - not going to be picky about this. Please don't let this block the merge

Reply:
Sorry, missed replying in proper detail!

The primary purpose of the test is to assert that even though we are repeatedly introducing a "selection bias" (only moving particles that are in some specified region of the box), the average number of ideal gas particles in that region does not decrease over time. Naively, the average number of particles in the resampled region would eventually decrease to 0.

This secondary assertion is intended to enforce that the primary assertion is not passing trivially. (One way to trivially satisfy an assertion of "no drift in this specific observable" is to return a constant trajectory -- "no drift in any observable." Want the test to rule out this possibility, since you could accidentally be in this situation if, for example, the test accidentally used an overly large step size, so that 0% of proposed moves were being accepted.) 

Another, stronger indication that the primary assertion is not passing trivially is that the primary assertion fails if the same experiment is performed using all the same sampler settings etc., just commenting out the restraint (https://github.com/proteneer/timemachine/blob/b1deb66cc5ae436a6612c44d9fd5637bcf941471/tests/test_local_move.py#L179-L181 ).

I'll also make this secondary assertion stronger for completeness (assert that all particles in the resampled region have moved at least a little).

Reply:
Applied in https://github.com/proteneer/timemachine/pull/785/commits/8709b31af63867a49a3365e3a0570cb51f735ba6